usr/src/uts/common/fs/zfs/arc.c

```diff
@@ -3054,6 +3662,164 @@ arc_hdr_realloc(arc_buf_hdr_t *hdr, kmem_cache_t *old, kmem_cache_t *new)
+		nsize = sizeof (hdr->b_crypt_hdr);
+		ocache = hdr_full_cache;
+		osize = HDR_FULL_SIZE;
+	} else {
+		ncache = hdr_full_cache;
+		nsize = HDR_FULL_SIZE;
+		ocache = hdr_full_crypt_cache;
+		osize = sizeof (hdr->b_crypt_hdr);
+	}
+
+	nhdr = kmem_cache_alloc(ncache, KM_PUSHPAGE);
+
+	/*
+	 * Copy all members that aren't locks or condvars to the new header.
+	 * No lists are pointing to us (as we asserted above), so we don't
+	 * need to worry about the list nodes.
+	 */
+	nhdr->b_dva = hdr->b_dva;
+	nhdr->b_birth = hdr->b_birth;
+	nhdr->b_type = hdr->b_type;
+	nhdr->b_flags = hdr->b_flags;
+	nhdr->b_psize = hdr->b_psize;
+	nhdr->b_lsize = hdr->b_lsize;
+	nhdr->b_spa = hdr->b_spa;
+	nhdr->b_l2hdr.b_dev = hdr->b_l2hdr.b_dev;
+	nhdr->b_l2hdr.b_daddr = hdr->b_l2hdr.b_daddr;
+	nhdr->b_l1hdr.b_freeze_cksum = hdr->b_l1hdr.b_freeze_cksum;
+	nhdr->b_l1hdr.b_bufcnt = hdr->b_l1hdr.b_bufcnt;
+	nhdr->b_l1hdr.b_byteswap = hdr->b_l1hdr.b_byteswap;
+	nhdr->b_l1hdr.b_state = hdr->b_l1hdr.b_state;
+	nhdr->b_l1hdr.b_arc_access = hdr->b_l1hdr.b_arc_access;
+	nhdr->b_l1hdr.b_acb = hdr->b_l1hdr.b_acb;
+	nhdr->b_l1hdr.b_pabd = hdr->b_l1hdr.b_pabd;
+#ifdef ZFS_DEBUG
+	if (hdr->b_l1hdr.b_thawed != NULL) {
+		nhdr->b_l1hdr.b_thawed = hdr->b_l1hdr.b_thawed;
+		hdr->b_l1hdr.b_thawed = NULL;
+	}
+#endif
+
+	/*
+	 * This refcount_add() exists only to ensure that the individual
+	 * arc buffers always point to a header that is referenced, avoiding
+	 * a small race condition that could trigger ASSERTs.
+	 */
+	(void) refcount_add(&nhdr->b_l1hdr.b_refcnt, FTAG);
+	nhdr->b_l1hdr.b_buf = hdr->b_l1hdr.b_buf;
+	for (buf = nhdr->b_l1hdr.b_buf; buf != NULL; buf = buf->b_next) {
+		mutex_enter(&buf->b_evict_lock);
+		buf->b_hdr = nhdr;
+		mutex_exit(&buf->b_evict_lock);
+	}
+	refcount_transfer(&nhdr->b_l1hdr.b_refcnt, &hdr->b_l1hdr.b_refcnt);
+	(void) refcount_remove(&nhdr->b_l1hdr.b_refcnt, FTAG);
+	ASSERT0(refcount_count(&hdr->b_l1hdr.b_refcnt));
+
+	if (need_crypt) {
+		arc_hdr_set_flags(nhdr, ARC_FLAG_PROTECTED);
+	} else {
+		arc_hdr_clear_flags(nhdr, ARC_FLAG_PROTECTED);
+	}
+
+	/* unset all members of the original hdr */
+	bzero(&hdr->b_dva, sizeof (dva_t));
+	hdr->b_birth = 0;
+	hdr->b_type = ARC_BUFC_INVALID;
+	hdr->b_flags = 0;
+	hdr->b_psize = 0;
+	hdr->b_lsize = 0;
+	hdr->b_spa = 0;
+	hdr->b_l2hdr.b_dev = NULL;
+	hdr->b_l2hdr.b_daddr = 0;
+	hdr->b_l1hdr.b_freeze_cksum = NULL;
+	hdr->b_l1hdr.b_buf = NULL;
+	hdr->b_l1hdr.b_bufcnt = 0;
+	hdr->b_l1hdr.b_byteswap = 0;
+	hdr->b_l1hdr.b_state = NULL;
+	hdr->b_l1hdr.b_arc_access = 0;
+	hdr->b_l1hdr.b_acb = NULL;
+	hdr->b_l1hdr.b_pabd = NULL;
+
+	if (ocache == hdr_full_crypt_cache) {
+		ASSERT(!HDR_HAS_RABD(hdr));
+		hdr->b_crypt_hdr.b_ot = DMU_OT_NONE;
+		hdr->b_crypt_hdr.b_ebufcnt = 0;
+		hdr->b_crypt_hdr.b_dsobj = 0;
+		bzero(hdr->b_crypt_hdr.b_salt, ZIO_DATA_SALT_LEN);
+		bzero(hdr->b_crypt_hdr.b_iv, ZIO_DATA_IV_LEN);
+		bzero(hdr->b_crypt_hdr.b_mac, ZIO_DATA_MAC_LEN);
+	}
+
+	arc_space_consume(nsize, ARC_SPACE_HDRS);
```

Comment:
The kmem_cache constructors (`hdr_full_cons` and `hdr_full_crypt_cons` and destructors (`hdr_full_dest` and `hdr_full_crypt_dest`) they already call arc_space_{consume,return} on ARC_SPACE_HDRS. Won't these end up double counting things and mess up the arc totals? I didn't see any other kmem_cache calls to either cache also doing an arc_space_* call.

Reply:
I tested removing these two `arc_space_consume()`calls and have been unable to replicate the ASSERT failures afterwards (I also no longer see any `aggsum_t`s go negative).  Combined with the `_cons` and `_dest` functions, it makes me feel fairly confident that these statements are incorrect and should be removed.

Reply:
Just that call to `arc_space_consume()` ?  You said "statementS" though?

Reply:
@GernotS Would you be willing to retry with the latest updates to see if they fix your crashes and corruption issues?  With the latest update (thanks @lundman for that), I've not been able to reproduce any other problems, so I'd be interested to see if you are able to.

Reply:
still crashing, but different:

panic[cpu2]/thread=fffffe00406bac20:
Invalid dnode block MAC


fffffe00406ba320 zfs:zfs_nfsshare_inited+37d81033 ()
fffffe00406ba430 zfs:dbuf_sync_leaf+2d8 ()
fffffe00406ba480 zfs:dbuf_sync_list+bb ()
fffffe00406ba4d0 zfs:dbuf_sync_indirect+b7 ()
fffffe00406ba520 zfs:dbuf_sync_list+ac ()
fffffe00406ba570 zfs:dbuf_sync_indirect+b7 ()
fffffe00406ba5c0 zfs:dbuf_sync_list+ac ()
fffffe00406ba610 zfs:dbuf_sync_indirect+b7 ()
fffffe00406ba660 zfs:dbuf_sync_list+ac ()
fffffe00406ba6b0 zfs:dbuf_sync_indirect+b7 ()
fffffe00406ba700 zfs:dbuf_sync_list+ac ()
fffffe00406ba750 zfs:dbuf_sync_indirect+b7 ()
fffffe00406ba7a0 zfs:dbuf_sync_list+ac ()
fffffe00406ba820 zfs:dnode_sync+259 ()
fffffe00406ba950 zfs:dmu_objset_sync+185 ()
fffffe00406ba9c0 zfs:dsl_dataset_sync+85 ()
fffffe00406baa60 zfs:dsl_pool_sync+b8 ()
fffffe00406bab40 zfs:spa_sync+477 ()
fffffe00406bac00 zfs:txg_sync_thread+247 ()
fffffe00406bac10 unix:thread_start+8 ()

Reply:
Can you share any additional details on the fs?  Was this just doing a zfs send/recv (though there it looks like it was trying to share out a fs via nfs)?  Do you know if any of the files on that fs had ACLs, extended attributes, or anything less common?

Reply:
Was this on a DEBUG kernel or non-DEBUG?  If it was a non-DEBUG kernel, would you be willing to try a DEBUG kernel?  The additional checks can sometimes spot a problem closer to the source.  Also, do you know what commit (or I guess what the last commit was) you applied the PR to?

Reply:
trying debug now, latest bits